transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollSocketChannel.java

```diff
@@ -110,6 +114,33 @@ protected AbstractEpollUnsafe newUnsafe() {
         return new EpollSocketChannelUnsafe();
     }
 
+    @Override
+    boolean doConnect0(SocketAddress remote) throws Exception {
+        if (Native.IS_SUPPORTING_TCP_FASTOPEN && config.isTcpFastOpenConnect()) {
+            pendingFastOpenWrite = true;
+            return true;
+        }
+        return super.doConnect0(remote);
+    }
+
+    @Override
+    int writeBytes(ChannelOutboundBuffer in, ByteBuf buf) throws Exception {
+        if (pendingFastOpenWrite && buf.isReadable()) {
+            pendingFastOpenWrite = false;
+            long localFlushedAmount = doWriteOrSendBytes(buf, remoteAddress(), true);
+            if (localFlushedAmount > 0) {
```

Comment:
This.... kind of sucks for testing. The client cookies are external state stored in the system, I did encounter the EINPROGRESS state until after I reset the machine. DNS cache is at least... idempotent, this seems to be some next-level machine state concerns.